File: src/main/java/ConfigurationManager.java
public class ConfigurationManager {
    // Volatile keyword ensures that multiple threads handle the uniqueInstance variable correctly when it is being initialized to the Singleton instance
    private static volatile ConfigurationManager uniqueInstance;

    // Configuration settings
    private String databaseURL;
    private String apiKey;

    // Private constructor prevents instantiation from other classes
    private ConfigurationManager() { }

    // Double-checked locking to reduce the use of synchronization in getInstance()
    public static ConfigurationManager getInstance() {
        if (uniqueInstance == null) {
            synchronized (ConfigurationManager.class) {
                if (uniqueInstance == null) {
                    uniqueInstance = new ConfigurationManager();
                }
            }
        }
        return uniqueInstance;
    }

    // Set and get configuration entries
    public void setDatabaseURL(String databaseURL) {
        this.databaseURL = databaseURL;
    }

    public String getDatabaseURL() {
        return databaseURL;
    }

    public void setApiKey(String apiKey) {
        this.apiKey = apiKey;
    }

    public String getApiKey() {
        return apiKey;
    }
}

File: src/ConfigurationManager.ts
class ConfigurationManager {
  private static uniqueInstance: ConfigurationManager | null = null;
  private databaseURL: string | null = null;
  private apiKey: string | null = null;

  // Private constructor to prevent instantiation
  private constructor() {}

  // Static method to get the single instance of the class
  public static getInstance(): ConfigurationManager {
    if (!ConfigurationManager.uniqueInstance) {
      ConfigurationManager.uniqueInstance = new ConfigurationManager();
    }
    return ConfigurationManager.uniqueInstance;
  }

  // Setters and getters for configuration settings
  public setDatabaseURL(databaseURL: string | null): void {
    this.databaseURL = databaseURL;
  }

  public getDatabaseURL(): string | null {
    return this.databaseURL;
  }

  public setApiKey(apiKey: string | null): void {
    this.apiKey = apiKey;
  }

  public getApiKey(): string | null {
    return this.apiKey;
  }
}

export default ConfigurationManager;
